### DISCUSSION TITLE
Image Handling

### DISCUSSION THREAD

**Discussion by torvista** [zencart/zencart#6380] [Ideas]
Do any sites continue to use the zc-inbuilt-three-images system/do **not** install Image Handler?

Do other carts use a similar method to Image Handler/is there a better method to implement?

**Comment by zenexpert** [comment]
Extending to torvista's question, is there a specific reason why we're still using the additional images approach through matching filenames? What would be the downsides of creating a separate table for additional images and storing them in the database just like the main image? 

Scanning the images directory and relying on preg_match to find matching filenames can become really inefficient, especially when there are many images in the directory, it can add quite some load to the CPU and disk I/O, and can easily return mistakes because files weren't named exactly according to requirements (which the end user doesn't really want to bother with or simply doesn't understand).

If switched to database approach, additional images could easily be added directly from the admin product page (making the whole process simpler), images could easily have different names and even extensions, queries could be cached etc etc.

**Comment by torvista** [comment]
I have nothing in the root /images apart from zc core images.

I DO have 29,000 images in 280 folders and don't see a resources problem.

Every.single.set of images being 
main image name: sku_product_name.jpg:...added with the usual admin product edit
additional images: sku_product_name_some_description.jpg....get found automatically/don't need defining anywhere, whatsoever.

I can't imagine anything simpler nor easier to manage: organise your products in folders per product or product family and use logical names.

Hence I don't get why the ZC default image system has not been supplanted by Image Handler years ago.

**Reply by zenexpert** [reply]
I couldn't agree more and this is something I **try** to get people to do, but at the end of the day, this depends on YOU being well organized. Many people are not, or have found themselves in a situation where nobody suggested this before and now they have thousands of images sitting in the root /images directory...
Your approach would still be excellent and easier to maintain, of course, but I don't see it colliding with the database approach - they would actually work beautiful hand in hand.

My argument to switch to DB approach would be to simplify things for the average user, not someone who actually understands file naming conventions. You'd be surprised how many people struggle with incorrect filenames, or wonder why image_01.png is not recognized as an additional image for image1.jpg. There's also supplier feeds which would be easier to implement when you get a list of additional images and you don't need to rename them, you just get to use the data provided. What makes things even worse with such integrations is when supplier adds 2 new additional images to the feed and you've already renamed the previous ones so now you don't know which one is new and which one already exists... 

My question was actually driven by an idea to change the way images are added to the product. Imagine a drag-and-drop field on the admin product page where you simply drop 10 images and choose the main one. Isn't that better than the current way of doing it? I would be happy to come up with a solution for this, but I'm not even gonna start with it if the idea to use a DB approach is not welcome.

Besides, what other platform nowadays uses directory scanning to try and detect additional images?

**Comment by barco57** [comment]
> Every.single.set of images being
main image name: sku_product_name.jpg:...added with the usual admin product edit
additional images: sku_product_name_some_description.jpg....get found automatically/don't need defining anywhere, whatsoever.

I am so glad you are organized...
Imagine having 40 old lady clients and trying to get them to do this. filenames camelcased, spaces in the file names, all taking their pictures on their new Iphones with the filename given by the phone, etc... Image handler is required to deal with the image naming convention for Zen Cart or chaos ensues.

**Comment by barco57** [comment]
> Imagine a drag-and-drop field on the admin product page where you simply drop 10 images and choose the main one. Isn't that better than the current way of doing it? 

This is exactly the conversation I have had with Lat9

**Reply by zenexpert** [reply]
And what was the outcome of the convo? Was it here on github or on the forums, any chance for me to catch up and see thoughts, maybe ups and downs?

I know the same could in theory be done with the current image system and I was on the verge of creating the plugin, but think a DB approach would be simpler and would need to be done first. It's not the file naming conventions that bother me, it's the directory scanning which uses more resources that a DB lookup, plus can easily return incorrect results. 

Anyway - I'm hoping to hear the **downsides** of using the DB approach, OR why the directory scanning is better. The implementation itself would be rather simple and I'm happy to provide it (core only, not Image Handler), but I don't want to waste time on it if it will never be adopted or if it could potentially cause more problems than it's worth.

**Reply by barco57** [reply]
It was a voice conversation on skype, nothing concrete about it was fleshed out. It was mostly me throwing out ideas for image handling interface changes on the product page in the admin. Never got as far as the details

**Reply by lat9** [reply]
>
> ...
> Anyway - I'm hoping to hear the **downsides** of using the DB approach, OR why the directory scanning is better. The implementation itself would be rather simple and I'm happy to provide it (core only, not Image Handler), but I don't want to waste time on it if it will never be adopted or if it could potentially cause more problems than it's worth.

Downsides:
- **Database bloat**.  I've got a client with 24,500 products, 50% of those have at least one additional image.  Assuming that each database record includes an auto-increment, an int(11) products_id and a varchar(255) image name, that's an additional 9.7Mb of storage required (over and above the images themselves).
- **Initialization time**.  For an upgrade, I'm guessing roughly 45-60 minutes for the referenced site's database to be synchronized with the currently-present images.
- **Overall synchronization**.  Since the database has no control over what happens in the file-system (and vice-versa), keeping that database table updated with file-system 'reality' could be a challenge.  We've had enough reports of sites where the product's (single) image name gets out-of-sync with its file-system presence.

**Reply by zenexpert** [reply]
> Downsides:
> 
>     * **Database bloat**.  I've got a client with 24,500 products, 50% of those have at least one additional image.  Assuming that each database record includes an auto-increment, an int(11) products_id and a varchar(255) image name, that's an additional 9.7Mb of storage required (over and above the images themselves).
Am I missing something obvious here? Why is 10 MB an issue nowadays when disk space is measured in GB or is unmetered on most shared hosts? 
Also, I'd say the calculation is a bit off - 20.000 records, with id, products_id and image_name (which would usually be smaller, but I'll go wild and say 100 characters), I'm getting roughly 2.3 MB, or with indexes 3.5 MB.

>     * **Initialization time**.  For an upgrade, I'm guessing roughly 45-60 minutes for the referenced site's database to be synchronized with the currently-present images.
Not sure I understand this. Are you talking about the site with 24k products and saying it would take 45-60 minutes just to populate the current additional images? My estimate for such a case would be 3-5 minutes.

>     * **Overall synchronization**.  Since the database has no control over what happens in the file-system (and vice-versa), keeping that database table updated with file-system 'reality' could be a challenge.  We've had enough reports of sites where the product's (single) image name gets out-of-sync with its file-system presence.
I can't say I agree with this as an argument because it's an issue now as well. If you MANUALLY rename a file and something breaks, well... If I rename configure.php to settings.php and come crying about it, will you say it's a system fault or that I'm just brain-dead? ;)

What about a side-by-side comparison, all the pros and cons of the DB driven approach? WooCommerce, Magento, OpenCart, Prestashop, Shopify - they all use DB-driven approach. osC also switched to DB-driven approach... I'm just wondering why we can't do it

**Comment by g2ktcf** [comment]
this is the major difficulty I have with trying to convince other users to
adopt ZC.  What Micheal refers to above would be awesome.

**Comment by torvista** [comment]
Since I don't support any other Zen Cart users I can only read in bewilderment at the level of incompetence that is being experienced/assumed.
If a potential shop owner or in fact any human old enough to vote is incapable of understanding a single sentence that explains how images should be named, that is scary. Still, that would explain the current state of the world.

I evidently don't have any idea of how low we have to go. 

However the aspect of dealing with supplier feeds is indeed very relevant.

**Reply by barco57** [reply]
> bewilderment at the level of incompetence that is being experienced/assumed.

all I do is support zc hosting clients and most of them were given instructions, written out, on how to manage images. Unfortunately it was more than 3 sentences and might have worked when they first populated the site with the first 50 products... then time went by and they just upload the image they took with their Iphone, with no resizing or optimization and wonder why it didn't work cause IH said no to the image being over 5MB, and I have to explain again image management when they run out of disk space. Most of the time they wonder why ZC doesn't handle all the fiddly details for them.

**Reply by torvista** [reply]
You've mentioned using IH to add additional images...I've seen that in the IH admin and must admit I've no idea about that...I've never done anything manual to add more images apart from upload them with the correct name.

The idea of being able to use any name just seems like chaos to me...also means that images for a product are no longer grouped together if you are browsing a folder.

However my lack of experience with other shop owners inabilities does not prevent me from accepting that you've seen a need and the benefits, and I trust that yourself and those who implement any new system of image handling know what they are doing.

**Reply by zenexpert** [reply]
You would be surprised to see what goes on out in the wild... 
I have clients who just put EVERYTHING in the main images/ directory. Just checked one of them - 29,153 images in the directory (excluding subdirectories). 
There's people who stare at you when you tell them to use FTP - like what's that? 

I agree using any name not only seems like chaos, it in fact IS chaos, but as Mike already mentioned - people will take a photo with their phone and upload it straight to the website. Rename it to match the main image? Heck, I'd struggle to rename a file on my phone, now imagine a user who just wants it to work and couldn't care less about the technicalities... They can do it on their other website that's using shopify or wp, why can't they do it here?

All that aside and back to the main reason where this idea actually originates from - all other platforms are doing it this way. If Zen Cart wants to take the path of being an ecommerce platform for developers, then it's not a problem and this idea can easily be binned. If it's meant for the end user who wants a simple and INTUITIVE way of managing their shop, then things need to be simplified to the max. There's sooooo many stupid-simple little things that could be improved and could help bring Zen Cart back to life, because atm I'd say it's on life support... This is/was being discussed in another thread so I won't reopen it here...

**Reply by torvista** [reply]
I understand your phone example, likewise.

Although I would not dream of uploading a raw phone image. I take a bunch of photos, upload to Dropbox to view on a proper screen, select the best and edit,  adjust, crop,  rename as necessary and save to a local dev folder that mirrors the production site. Guess that's too much for most people.

If just on the basis of giving the user what they think they need, if that helps the project, it's enough. If everyone else does it....

Unfortunately the occasional discussions of what needs to be done to breath life into ZC never lead to anything new like a plan, it's still ad-hoc tinkering.
